### DISCUSSION TITLE
Node classification with StellarGraph - Targets as dictionaries and overall pipeline

### DISCUSSION THREAD

**Discussion by amjass12** [stellargraph/stellargraph#1777] [General]
Hi all,

I am bringing this discussion from the discourse forum as I successfully built a GCN node classification model, but would like some clarification to some points, but also ask another question for a bigger aim that I have downstream. 

I have correlation matrices of various graphs. I build individual graphs using networkX with the aim of identifying communities within each graph. I would now like to classify nodes based on the ground truth for communities they belong to using a GCN in stellargraph. My code and specific questions are below: I really appreciate any input and thanks in advance!

My correlation matrix is pre-filtered to include nodes that have a value of 0.8 or above only. This produces an x by x matrix (for an example, call it a 457 x 457 matrix). 

a networkX graph is produced as follows:
```

import networkx as nx
import community
networkGraph14=nx.Graph(corr14).to_undirected() #corr14 being the correlation matrix of pearson correlation >0.8
comms = community.best_partition(networkGraph14) ##communities for networkGraph14 - generates ~60 communities
```

Note also - single nodes are excluded from this graph. only nodes that have at least one connection with another node are allowed (so 2 connected nodes minimum)
Note also, comms is a dictionary of keys node names (which are strings.. gene names) and values community member (either 1-60)

Now for the GCN graph to predict **communities** to which nodes exist: -- effectively, regenerate communities through learnt embeddings -- my aim later would be to take graphs that are generated from exactly the same genes in brand new data.. and predict membership to a specific graph community in training. A trained model of learnt embeddings is far more efficient that generating new graphs in networkx each time and attempting to find communities that are similar to another graph:
```

from stellargraph.mapper import FullBatchNodeGenerator

#generate node features
nodefeatures=corr14.iloc[corr14.index.isin(networkGraph14.nodes())]## take only nodes that were >0.8 from index

nodefeatures=nodefeatures.loc[:, nodefeatures.columns.isin(nodefeatures.index)]# from columns
-this generates node features that are 457 x 457. index and columns are gene names - values are the pearson correlation coefficients (>0.8 values only, so 457 x 457)

sGraph = StellarGraph.from_networkx(networkGraph14, node_features=nodefeatures)
print(sGraph.info())

StellarGraph: Undirected multigraph
 Nodes: 457, Edges: 3174

 Node types:
  default: [457]
    Features: float32 vector, length 457
    Edge types: default-default->default

 Edge types:
    default-default->default: [3174]
        Weights: range=[0.800066, 1], mean=0.872295, std=0.0648471
        Features: none
# create target data -- communities to learn
target=pd.DataFrame(comms.values(), index=comms.keys())

target.head
ENSMUSG00000002459   0
ENSMUSG00000025932   1
ENSMUSG00000041859   2
ENSMUSG00000026051   0
ENSMUSG00000025983   3
...                 ..
ENSMUSG00000074896  55
ENSMUSG00000062488  55
ENSMUSG00000024795   2
ENSMUSG00000012443   2
ENSMUSG00000025154   6

#split data
train_subjects, test_subjects = model_selection.train_test_split(
    target, train_size=420, test_size=None
)#stratify was removed as there communities of 2 or less exist other wise an error is thrown
val_subjects, test_subjects = model_selection.train_test_split(
    test_subjects, train_size=36, test_size=None
)#stratify was removed as there communities of 2 or less exist other wise an error is thrown

target_encoding = preprocessing.LabelBinarizer()

train_targets = target_encoding.fit_transform(train_subjects)
val_targets = target_encoding.transform(val_subjects)
test_targets = target_encoding.transform(test_subjects)

generator = FullBatchNodeGenerator(sGraph, method="gcn")
train_gen = generator.flow(train_subjects.index, train_targets)

gcn = GCN(
    layer_sizes=[64,32], activations=["relu", "relu"], generator=generator, dropout=0.5
)## I found that increasing neuron number improved accuracy in data


x_inp, x_out = gcn.in_out_tensors()

predictions = tf.keras.layers.Dense(units=train_targets.shape[1], activation="softmax")(x_out)

model = tf.keras.models.Model(x_inp, predictions)
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=["acc"],
)
val_gen = generator.flow(val_subjects.index, val_targets)

from tensorflow.keras.callbacks import EarlyStopping

es_callback = EarlyStopping(monitor="val_loss", patience=50, restore_best_weights=True)

history = model.fit(
    train_gen,
    epochs=200,
    validation_data=val_gen,
    verbose=1,
    shuffle=False, 
    callbacks=[es_callback]
)##training performance attached

#evaluate performance and downstream embedding tSNE etc

all_nodes = list(comms.keys())
all_gen = generator.flow(all_nodes)
all_predictions = model.predict(all_gen)

node_predictions = target_encoding.inverse_transform(all_predictions.squeeze())
df = pd.DataFrame({"Predicted": node_predictions, "True": list(comms.values())})

```

<img width="573" alt="Screenshot 2020-07-24 at 16 42 07" src="https://user-images.githubusercontent.com/33659783/88409269-a0062080-cdcc-11ea-99db-5804b68bef92.png">

The specific questions that I have are:

**during train/test split, because some communities are small- there should be no problem setting stratify to False right? 

Generating communities from a dictionary and in to a Pandas dataframe where node names are indexes and community membership are columns are ok? (i don't see why they shouldn't as it is essentially simply making a pandas dataframe from the dict object before creating targets and then one-hot encoding them)

the line `train_gen = generator.flow(train_subjects.index, train_targets)` ensures the correct nodes are identified from the graph structure?

early stopping does not come in to effect and the model fits for 200 epochs as the val_loss continues to decrease.. not a stellargraph question, but presumably this is normal behaviour for a graph net? -normally a keras trained mode for me converges relatively quickly (~30 epochs) -- additionally, the validation accuracy seems to plateu for many epochs before increasing again, is there an explanation for this? **EDIT- adjusting learning rate fixed this problem and data converge at ~26 epochs..** 

And finally, I would like to train a GCN to learn communities present in up to 50 graphs, all 50 representing genes from different ages etc.. Would you recommend a union of all graphs joined in to one bigger graph and then trained as above to learn the communities? node names from different graphs will be common in all graphs, but it is the community membership at different ages that could change... this should still be a feasible approach? **

I know this is a long post, but I appreciate any advice and help! thanks!!

**Comment by huonw** [comment]
> I am bringing this discussion from the discourse forum as I successfully built a GCN node classification model, but would like some clarification to some points, but also ask another question for a bigger aim that I have downstream.

Great work, and thanks for working with us during the migration!

(For future reference, the original thread: [original](https://community.stellargraph.io/t/converting-from-networkx-node-feature-error/153), [archive](https://web.archive.org/web/20200716034859/https://community.stellargraph.io/t/converting-from-networkx-node-feature-error/153).)

> -this generates node features that are 457 x 457. index and columns are gene names - values are the pearson correlation coefficients (>0.8 values only, so 457 x 457)

As noted in the previous discussion, the node features don't have to be N&times;N. I'd encourage you to investigate using `counts` as the features directly. Reusing the correlations means the node features capture similar information to the edges, and only encode information about a node relative to the other nodes, not "inherent" information about each node. One way to look at graph models like GCN, GAT or GraphSAGE is they 'upgrade' an MLP model (which works on single data points, that is, just node features) with connections to other data points (the edges). 

If the `counts` data might be used in a non-graph model, a graph model could use it for the node features. If not, then continue as you were. :)

One thing to be aware of is the node features will need to be the same size for every graph you put into the model.

> **during train/test split, because some communities are small- there should be no problem setting stratify to False right?

You risk having more randomness about the proportions of the number of examples from each community. In particular, if some communities are very small, there may be no training examples chosen for that community. I think this is part of what the exception about class size is suggesting; there's some communities with only one node, so one cannot have both a train example and a test example for these communities.

> Generating communities from a dictionary and in to a Pandas dataframe where node names are indexes and community membership are columns are ok? (i don't see why they shouldn't as it is essentially simply making a pandas dataframe from the dict object before creating targets and then one-hot encoding them)

I'm not sure what you mean by this; are you referring to the following?

```python
# create target data -- communities to learn
target=pd.DataFrame(comms.values(), index=comms.keys())
```

If so, yes that seems perfectly fine. Although, it might be very slightly cleaner using [`from_dict`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.from_dict.html): `pd.DataFrame.from_dict(comms, orient="index"`)`.

> the line train_gen = generator.flow(train_subjects.index, train_targets) ensures the correct nodes are identified from the graph structure?

I'm not quite sure what you mean by this? I'll guess at an answer, but please let me know and rephrase the question if it doesn't make sense.

The first parameter to [the `flow` method](https://stellargraph.readthedocs.io/en/stable/api.html#stellargraph.mapper.FullBatchNodeGenerator.flow) is the node IDs to include for the given data generator. For supervised training, this should be the nodes to use for training along and then the targets to used for that training.

> early stopping does not come in to effect and the model fits for 200 epochs as the val_loss continues to decrease.. not a stellargraph question, but presumably this is normal behaviour for a graph net? -normally a keras trained mode for me converges relatively quickly (~30 epochs) -- [...] EDIT- adjusting learning rate fixed this problem and data converge at ~26 epochs..

Great 👍 

> additionally, the validation accuracy seems to plateu for many epochs before increasing again, is there an explanation for this?

I suspect that it's because the validation set is quite small and so the accuracy is very quantised: with only 36 nodes, it's a noticable jump of 2.8% (1/36) when one of those nodes goes from being classified incorrectly to being classified correctly. For a plot from 0-1 of height 300 pixels, this translates into about 8 pixels between each of the possible levels for the validation accuracy. A larger validation set will result in smoother increases of the validation accuracy, if you need this (e.g. for early stopping).

> And finally, I would like to train a GCN to learn communities present in up to 50 graphs, all 50 representing genes from different ages etc.. Would you recommend a union of all graphs joined in to one bigger graph and then trained as above to learn the communities? node names from different graphs will be common in all graphs, but it is the community membership at different ages that could change... this should still be a feasible approach? **

Should the information/edges from one age directly influence the information in another age (the other option is that the model has learn all of the commonalities)? Are the number of communities in each the age the same? 

If the answer to both of these is 'yes', then building one big graph with all of the nodes connected by edges from different ages might be a good approach. If the edges are labelled with which age they come from (to create a heterogeneous graph), then RGCN can be used. The output of this model for each node would be a `number of ages` &times; `number of communities` matrix, where each row (of size `number of communities`) is a softmax vector representing the chosen community in that age. This should be able to be trained with categorical cross-entropy.

If the answer is no, then an appropriate approach would be treating these as separate graphs. This could be done by placing them all into one big graph, but with distinct node IDs/names, so there's no overlap. (For example, node X from age 123 could be called `"123-X"`.) Another approach would be an 'inductive' setting, where the different ages are kept as separate `StellarGraph`s, and fed into the model separately. This can be done by creating [a `ClusterNodeGenerator`](https://stellargraph.readthedocs.io/en/stable/api.html#stellargraph.mapper.ClusterNodeGenerator) for each graph (passing it into `GCN`) or by switching to an algorithm like [GraphSAGE](https://stellargraph.readthedocs.io/en/stable/api.html#stellargraph.layer.GraphSAGE) ([demo of inductive training](https://stellargraph.readthedocs.io/en/stable/demos/node-classification/graphsage-inductive-node-classification.html)). (If you want to train on more than one age/graph, it might be best to combine these approaches: combine all the training graphs into one big one, with separate nodes, and use `ClusterNodeGenerator`, but for prediction, feed each age/graph through the model separately.)

---

I've written done a lot of thoughts here, I hope they're enough to give you some ideas and get you started.

**Reply by amjass12** [reply]
Hi @huonw 

thank you so much as always for your detailed replies!

To your first point: I have read the documentation thoroughly and your explanations of the node features and now understand this parameter... 

So the `counts.transpose()` is a rows of features (nodes) and columns of samples (~1000 samples): The correlation matrix is then made directly from this: so the counts table looks like this:


counts.head()
               features (nodes)                                           sample 1 --- up to sample 1000
ENSMUSG00000065947                                         492.108000                            
ENSMUSG00000064363                                        9462.808000                            
ENSMUSG00000064367                                       18839.850000                            
ENSMUSG00000064368                                        1523.863000

when i run: `nodefeatures=normCountsscale1.iloc[normCountsscale1.index.isin(networkGraph14.nodes())]`

this all works fine -- 457 x 1000 matrix (of original counts)
but the downstream `model.fit` now now achieves a **maximum of ~0.1% acc and val_acc during training**. Presumably this is because the counts are the original floating points.

-when running the data as variance stabilising transformed values (range roughly between 3-20) the same **low** accuracy is achieved
- when i scaled the data above to between 0 and 1 (the original data not VST counts)- the model now returns to its normal high accuracy that it achieved as in my original post... 

Presumably this is fine? the data would be scaled to between 0 and 1 when running any other MLP so i am guessing this is ok. 
question then: the scaled data from the original counts (457 x1000 matrix) should be fine? and should not interfere with the edge data (which is the correlation values of between 0.8 and 1)? I want to make sure that although the metrics plot dont suggest overfitting, that there isn't any sorry of memorisation of the nodes during training because of the node values (scaled between 0 and 1 in the node features parameter)... 

Is the behaviour of low accuracy when inputting the high count values the same behaviour as if you were to pass this data without scaling it to a straightforward MLP?

>You risk having more randomness about the proportions of the number of examples from each community. In particular, if some communities are very small, there may be no training examples chosen for that community. I think this is part of what the exception about class size is suggesting; there's some communities with only one node, so one cannot have both a train example and a test example for these communities.

Perfectly clear-- i think however this is a caveat i might have to live with for now given the size of some communities... 

> I'm not sure what you mean by this; are you referring to the following?
Indeed thats what i meant and you answered my question perfectly, thanks!

The first parameter to the flow method is the node IDs to include for the given data generator. For supervised training, this should be the nodes to use for training along and then the targets to used for that training.

This also answers my question thank you! 

> I suspect that it's because the validation set is quite small and so the accuracy is very quantised: with only 36 nodes, it's a noticable jump of 2.8% (1/36) when one of those nodes goes from being classified incorrectly to being classified correctly. For a plot from 0-1 of height 300 pixels, this translates into about 8 pixels between each of the possible levels for the validation accuracy. A larger validation set will result in smoother increases of the validation accuracy, if you need this (e.g. for early stopping

Also very clear, thank you 👍 

To your last point:

>Should the information/edges from one age directly influence the information in another age (the other option is that the model has learn all of the commonalities)? Are the number of communities in each the age the same?
The edges do not necessarily need to influence the information in another age- I would ideally like to see the transition from age to another. so if community at age1 has connected nodes x-y-z.. I would like to see if at age2, maybe this community disbands and you might start to see this community consisting of nodes x-y. The number of communities in each age is **NOT** the same.

Presumably this is your first point about having a graph of all communities. My concern with this is that we might start to create edges connecting nodes from age 1 and age 5.. which is completely artificial and wouldn't exist under normal circumstances?

My original thinking was your second point about one bigger graph of disconnected graphs (ages) uniquely labelled to represent those ages- although your first point is really interesting. Maybe i have misunderstood something about how ages might be connected together whilst still maintaining age specific communities?

thank you so much again for your continued support and detailed replies!
